Use byteSwapX# in WORDS_BIGENDIAN

sgraf812 opened the pull request:
Spurred by @Bodigrim in https://github.com/haskell/happy/pull/280#discussion_r1651705321, let's try having a bit more efficient code on `WORD_BIGENDIAN` architectures.

Mostly uploading this for CI, hoping there will be a big-endian machine to test the new code path.

sgraf812 commented:
This seems to work as expected, presuming the MacOS build really tests the `WORDS_BIGENDIAN` code path. And I couldn't tell how it would work otherwise.

andreasabel commented:
Thanks!

Could we get a changelog entry for this?

sgraf812 commented:
Done. I say 3.5.2.0 in the CHANGELOG, but I'm inclined to think it's only a PATCH change really. You can always adjust later, I guess.

Bodigrim commented:
> This seems to work as expected, presuming the MacOS build really tests the `WORDS_BIGENDIAN` code path.

Unfortunately it does not, Apple hardware is little endian for the past 15 years or so. 

To test the change properly one has to setup an emulated job for `s390x` architecture (aka IBM mainframe). There are several examples of such CI setup around github.com/haskell, the most recent one I worked on is https://github.com/haskell/tar/blob/master/.github/workflows/emulated.yml.

sgraf812 commented:
Oh, wow... I somehow thought that iOS was big endian, but it actually isn't. Nevermind then.

Bodigrim commented on data/AlexTemplate.hs:
```diff
@@ -44,18 +44,15 @@ data AlexAddr = AlexA# Addr#
 #endif
-    (indexInt16OffAddr# arr off)
+#ifdef WORDS_BIGENDIAN
+  (GHC.Exts.word16ToInt16# (GHC.Exts.byteSwap16# (GHC.Exts.int16ToWord16#
```
Annoyingly [`byteSwap16#`](https://hackage.haskell.org/package/base-4.20.0.1/docs/GHC-Base.html#v:byteSwap16-35-) is `Word# - > Word#`, not `Word16# -> Word16#`, so this needs amendment.

Cf. https://gitlab.haskell.org/ghc/ghc/-/issues/22163.

sgraf812 commented:
Alright, I added an emulated CI pipeline as @Bodigrim  suggested. Not pretty, but does the job, and shouldn't break all to often given that `alex` is pretty stable.

For the job in https://github.com/sgraf812/alex/actions/runs/9912679649/job/27388017292 (569fb3f), I replaced `byteSwap` with garbage, so that job is expected to fail.

The job in https://github.com/sgraf812/alex/actions/runs/9912681641/job/27388023583 (36cc05e) does no such replacement and consequently is green.

No idea why the other CI jobs fail `cabal check`.

andreasabel commented:
> It really seems that all deps need to announce upper bounds now.

No, I think only `base` absolutely needs one now:
> Error: [missing-bounds-important] The dependency 'build-depends: base' does not specify an upper bound on the version number.

andreasabel commented:
I am fixing Haskell CI in this PR that we should merge first:
- https://github.com/haskell/alex/pull/261

andreasabel commented:
@Mergifyio rebase

andreasabel commented:
@sgraf812 : Do you want this change released now?

sgraf812 commented:
Thanks! No need for a release, I just wanted to test the code in https://github.com/haskell/happy/pull/280. Replicating the emulated-CI-without-cabal setup for the multi-component `happy` project is far too ambitious for my tastes.
